File: src/test/java/io/lsdconsulting/lsd/distributed/interceptor/captor/rabbit/AmqpHeaderRetrieverShould.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.rabbit;

import io.lsdconsulting.lsd.distributed.interceptor.captor.header.Obfuscator;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.amqp.core.Message;
import org.springframework.amqp.core.MessageProperties;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.mockito.AdditionalAnswers.returnsFirstArg;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class AmqpHeaderRetrieverShould {
    private final Obfuscator obfuscator = mock(Obfuscator.class);

    private final AmqpHeaderRetriever underTest = new AmqpHeaderRetriever(obfuscator);

    @BeforeEach
    void setup() {
        when(obfuscator.obfuscate(any())).then(returnsFirstArg());
    }

    @ParameterizedTest
    @MethodSource("provideMessageProperties")
    void retrieveHeadersFromMessages(MessageProperties messageProperties, int expectedSize) {
        final Message message = mock(Message.class);
        given(message.getMessageProperties()).willReturn(messageProperties);

        final Map<String, Collection<String>> result = underTest.retrieve(message);

        assertThat(result.keySet(), hasSize(expectedSize));
        for (String headerName : messageProperties.getHeaders().keySet()) {
            String header = messageProperties.getHeader(headerName);
            assertThat(result.get(headerName), contains(header));
        }
    }

    private static Stream<Arguments> provideMessageProperties() {
        MessageProperties mp1Value = new MessageProperties();
        mp1Value.setHeader("name", "value");
        MessageProperties mp2Values = new MessageProperties();
        mp2Values.setHeader("name1", "value1");
        mp2Values.setHeader("name2", "value2");
        return Stream.of(
                Arguments.of(mp1Value, 1),
                Arguments.of(mp2Values, 2),
                Arguments.of(new MessageProperties(), 0)
        );
    }

    @Test
    public void handleHeadersWithNoValuesFromMessage() {
        final MessageProperties messageProperties = new MessageProperties();
        messageProperties.setHeader("name", null);
        final Message message = mock(Message.class);
        given(message.getMessageProperties()).willReturn(messageProperties);

        final Map<String, Collection<String>> headers = underTest.retrieve(message);

        assertThat(headers.keySet(), hasSize(1));
        assertThat(headers.get("name"), is(List.of()));
    }
}

File: src/test/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/rabbit/AmqpHeaderRetrieverShould.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.rabbit

import io.lsdconsulting.lsd.distributed.interceptor.captor.header.Obfuscator
import io.mockk.every
import io.mockk.mockk
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.amqp.core.Message
import org.springframework.amqp.core.MessageProperties
import java.util.stream.Stream

internal class AmqpHeaderRetrieverShould {
    private val obfuscator = mockk<Obfuscator>()
    private val underTest = AmqpHeaderRetriever(obfuscator)

    @BeforeEach
    fun setup() {
        every { obfuscator.obfuscate(any()) } answers { firstArg() }
    }

    @ParameterizedTest
    @MethodSource("provideMessageProperties")
    fun retrieveHeadersFromMessages(messageProperties: MessageProperties, expectedSize: Int) {
        val message = mockk<Message>()
        every { message.messageProperties } returns messageProperties

        val result = underTest.retrieve(message)

        assertThat<Set<String>>(result.keys, hasSize(expectedSize))
        for (headerName in messageProperties.headers.keys) {
            val header = messageProperties.getHeader<String>(headerName)
            assertThat(result[headerName], contains(header))
        }
    }

    @Test
    fun handleHeadersWithNoValuesFromMessage() {
        val messageProperties = MessageProperties()
        messageProperties.setHeader("name", null)
        val message = mockk<Message>()
        every { message.messageProperties } returns messageProperties
        val headers = underTest.retrieve(message)

        assertThat<Set<String>>(headers.keys, hasSize(1))
        assertThat(headers["name"], `is`(listOf<Any>()))
    }

    companion object {
        @JvmStatic
        private fun provideMessageProperties(): Stream<Arguments> {
            val mp1Value = MessageProperties()
            mp1Value.setHeader("name", "value")
            val mp2Values = MessageProperties()
            mp2Values.setHeader("name1", "value1")
            mp2Values.setHeader("name2", "value2")
            return Stream.of(
                Arguments.of(mp1Value, 1),
                Arguments.of(mp2Values, 2),
                Arguments.of(MessageProperties(), 0)
            )
        }
    }
}
